### DISCUSSION TITLE
Example OSSL_DECODER to decode the ASN.1 encoded keys.

### DISCUSSION THREAD

**Discussion by stefbon** [openssl/openssl#21879] [Q&A] [answered]
Hi,

I'm looking for an example howto use the OSSL_DECODER to decode data with key material (private and public) from ASN.1.

The manpages suggest to use OSSL_DECODER. I've written a small program which walks every DECODER, and prints all names available per DECODER. I see DER is printed, (and ASN.1 is something DER related as far as I know), but then?

S. Bon

**Comment by t8m** [comment, marked as answer]
Please look at https://github.com/openssl/openssl/blob/master/demos/encrypt/rsa_encrypt.c

The get_key() function uses decoders to decode an RSA key from data in memory. OSSL_DECODER_from_bio() decodes keys from a BIO - which can be a file for example. You can also use NULL as the keytype argument of OSSL_DECODER_CTX_new_for_pkey() which allows decoding any key type.

**Reply by stefbon** [reply]
Thans for the answer. I will look into it.

**Comment by slontis** [comment]
There are also examples in doc/man3/OSSL_DECODER_from_bio.pod

**Reply by stefbon** [reply]
Thanks. It's very powerfull. Just give the decoder the right parameters and data, and it will do the job. Great. First I have to test this anyway.

**Comment by levitte** [comment]
A bit of information for you that I hope can clear some confusion:

- ASN.1 is a language to define structures and to specify data for those structures, i.e. it's the stuff you can read in standards like RFCs
- DER is a binary encoding format for data that are structured according to specifications written in ASN.1.  There is more than one encoding format, but DER seems to be the most popular and long lived.

**Reply by stefbon** [reply]
I know how ANS.1 works, written my own parser. See: 

https://github.com/stefbon/OSNS/tree/main/src/ssh/pk

Thanks anyway.
Stef

**Comment by stefbon** [comment]
I want to add that I've written an example to read a key from the openssh files like id_rsa and id_rsa.pub.
See:

https://github.com/stefbon/EVP_readkey_openssh

Note: only rsa works for now.
And the private key was easy (it uses the PEM format), the public key was not: openssh uses an own format.
